### ISSUE TITLE
Refact: Persistence of class PyTorchMLP

### ISSUE THREAD

**Issue by detlefarend** [fhswf/MLPro#673]
**Motivation**
Class PyTorchMLP can not be pickled. That makes it necessary to implement a specific load/save strategy. The underlying prerequisites are worked out in #664.

**Benefit**
Objects of this type can be saved/loaded.

**Are the proposed changes backward compatible?** 
- [ ] Yes
- [X] No

**Related development objects**
Class PyTorchMLP

**Related issues**
#664 

**Branch**
persistence

**Comment by detlefarend** [comment]
Hi @steveyuwono, I prepared a new class Persistent in branch persistence. It replaces classes Lodable, Saveable, LoadSave.

It uses builtin custom features of package pickle/dill.

I designed a small systematics based on own custom methods to deal with it as easy as possible.

The story goes like this:

If you pickle something then a standard method __getstate__() is called. This method returns the data structure self as a dictionary. Before returning this dictionary an mlpro custom method _reduce_state() is called. Here you need to do two things:
- remove everthing from the dict that can not be pickled
- save exactly those pices of unpickleable stuff in a suitable but different way. To this regard, a destinstion path and a filename without suffix (=filename stub) is handed over to the custom method. In case of your PyTorch components you can store them using the PyTorch methods, the filename stub and a specific suffix.

For the opposite direction (load) pickle provides the standard method __setstate__() to complete self from external data sources. The related mlpro custom method is _complete_state(). The systematics is the opposite of pickling something. You will understand if you see the code and howto_bf_005_persistence.py

If you have time and fun you could start implementing these custom methods in your MLP class. 

I need to refactor Scenario, gym and pettingzoo wrappers and so on. Agent, MultiAgent, EnvBase, ... need to be reviewed/refactored as well. Lets stay synched about the progress. And as always: a small zoom call can avoid wasting our worthful time;)

**Comment by steveyuwono** [comment]
> * remove everthing from the dict that can not be pickled

Hi @detlefarend , thank you for letting me know and for your explanation. I will have a look at it later today or this weekend. Then, I will enhance the PyTorch-related pickling issues with the custom method, which I believe will definitely work. If I have any difficulties, then I will reach you via slack.

Additionally, I have some issues while running our MLPro's RL training in the long run for the company project. I simulated RL training over 1,000,000 cycles with SB3 policy and ±80 states. With these settings, the generated storing file has reached 3,45GB, where 1.23GB is for storing the states in CSV files and 2.09GB is for storing the scenario. I believe this is insanely large. After simulating, here are some findings:

- Should we also introduce p_save_scenario, which is similar to p_collect_states, p_collect_rewards, etc.? Hence, the user can have an option to avoid saving the scenario and save memory, if they think it is unnecessary.
- Then, I tested by extending the cycle limit to 5,000,000 cycles. However, the pickling process ends up with an error due to excessive memory usage. I have enough ram, so I am unsure if pickling with that amount of memory can cause this issue.
- It takes too much time to perform data loading from CSV files and data plotting of a large data set (>1,000,000 cycles x 5 features) because of the looping, where we extract the rows one by one (or one cycle after another) and store it in MLPro's internal memory (self.memory_dict). This could also be an issue for us in the data stream tasks if dealing with external large datasets from another file. Reference: https://github.com/fhswf/MLPro/blob/26aa5bd3b442d7b60d622b1ce60e6766678caff7/src/mlpro/bf/data.py#L166-L177

I will also have a look at it deeper and try to find a better solution.

**Comment by detlefarend** [comment]
Hi @steveyuwono, regarding DataStoring and following your thoughts I propose the following: 
- we exclude it from pickling the scenario
- Option: we associate a DS class with a csv file
- Option: we don't buffer it in the memory but persist the data directly in the associated csv file. That could be done per line or by predefined block size
I think there is already an issue for it on hold. I'll take a look...

**Comment by steveyuwono** [comment]
> Hi @steveyuwono, regarding DataStoring and following your thoughts I propose the following:
> 
> * we exclude it from pickling the scenario
> * Option: we associate a DS class with a csv file
> * Option: we don't buffer it in the memory but persist the data directly in the associated csv file. That could be done per line or by predefined block size
>   I think there is already an issue for it on hold. I'll take a look...

Shall we exclude or provide an option of pickling the scenario from our base class?

**Comment by detlefarend** [comment]
Sorry, I didn't answer your question. I think, training without saving the scenario makes no sense. To reduce the size of a scenario we should exclude internal attributes _ds_states, _ds_actions, _ds_rewards from pickling. They are stored separately by the training class itself. Anyhow, let's discuss and enhance the features of class DataStoring to be able to limit it's data volume in memory and file...

**Comment by steveyuwono** [comment]
Hi @detlefarend , I have updated the PyTorchMLP class by adding _complete_state and _reduce_state to save and load the torch-related model separately using torch load/save functionalities. Additionally, I was adding _reduce_state in RLScenario to remove _ds_states, _ds_actions, and _ds_rewards from pickling, and then I just realized that you have already done it. That's why the unit test failed even after Rizky's updates. I would say that this Persistence class is really helpful.
